pkg/expapi/types.go

```diff
@@ -134,3 +136,40 @@ type HorizontalPodAutoscalerList struct {
 
 	Items []HorizontalPodAutoscaler `json:"items"`
 }
+
+type DerivedNodeMetrics struct {
+	api.TypeMeta   `json:",inline"`
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	NodeMetrics      MetricsWindows            `json:"nodemetrics"`
+	SystemContainers []DerivedContainerMetrics `json:"systemcontainers"`
+}
+
+type DerivedContainerMetrics struct {
+	Name    string         `json:"name"`
+	Windows MetricsWindows `json:"windows"`
+}
+
+type DerivedNodeMetricsList struct {
+	api.TypeMeta `json:",inline"`
+	api.ListMeta `json:"metadata,omitempty"`
+
+	Items []DerivedNodeMetrics `json:"items"`
+}
+
+// Last overlapping 10s, 1m, 1h and 1d as a start
+// Updated every 10s, so the 10s window is sequential and the rest are
+// rolling.
+type MetricsWindows map[time.Duration]DerivedMetrics
+
+type DerivedMetrics struct {
+	// End time of all the time windows in Metrics
+	EndTime util.Time `json:"endtime"`
```

Comment:
If EndTime is required, then it can stay `util.Time`, but you should have a
validation rule enforcing that it is non-zero.  If it is not required, and
the user's intent that they didn't set it is preserved, it would be
*util.Time.

On Fri, Aug 21, 2015 at 4:54 PM, Daniel Martí <notifications@github.com>
wrote:

> In pkg/expapi/types.go
> <https://github.com/kubernetes/kubernetes/pull/12952#discussion_r37677203>
> :
>
> > +
> > +type DerivedNodeMetricsList struct {
> > +	api.TypeMeta `json:",inline"`
> > +	api.ListMeta `json:"metadata,omitempty"`
> > +
> > +	Items []DerivedNodeMetrics `json:"items"`
> > +}
> > +
> > +// Last overlapping 10s, 1m, 1h and 1d as a start
> > +// Updated every 10s, so the 10s window is sequential and the rest are
> > +// rolling.
> > +type MetricsWindows map[time.Duration]DerivedMetrics
> > +
> > +type DerivedMetrics struct {
> > +	// End time of all the time windows in Metrics
> > +	EndTime util.Time `json:"endtime"`
>
> I don't think I follow. Are you saying that it could be present in
> pkg/expapi/v1/types.go but not in pkg/expapi/types.go?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12952/files#r37677203>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift